### ISSUE TITLE
Unexpected node: com.facebook.presto.sql.planner.plan.LateralJoinNode

### ISSUE THREAD

**Issue by georgewfraser** [prestodb/presto#8421]
Presto fails with the message `Unexpected node: com.facebook.presto.sql.planner.plan.LateralJoinNode` on a query based on TPCDS query 1:

```
WITH customer_total_return 
     AS (SELECT sr_customer_sk     AS ctr_customer_sk, 
                sr_store_sk        AS ctr_store_sk, 
                Sum(sr_return_amt) AS ctr_total_return 
         FROM   store_returns, 
                date_dim 
         WHERE  sr_returned_date_sk = d_date_sk 
                AND d_year = 2001 
         GROUP  BY sr_customer_sk, 
                   sr_store_sk),
high_return AS (
    SELECT ctr_store_sk, Avg(ctr_total_return) * 1.2 AS return_limit
    FROM   customer_total_return ctr2 
    GROUP BY ctr_store_sk
)
SELECT c_customer_id 
FROM   customer_total_return ctr1, 
       store, 
       customer 
WHERE  ctr1.ctr_total_return > (SELECT return_limit
                                FROM high_return
                                WHERE  ctr1.ctr_store_sk = high_return.ctr_store_sk) 
       AND s_store_sk = ctr1.ctr_store_sk 
       AND s_state = 'TN' 
       AND ctr1.ctr_customer_sk = c_customer_sk 
ORDER  BY c_customer_id
LIMIT 100;
```

It seems that this subquery is the problem:

```
(SELECT return_limit
FROM high_return
WHERE  ctr1.ctr_store_sk = high_return.ctr_store_sk)
```
You should be able to create an appropriate schema by running this in Hive:

```
create external table tpcds_parquet.customer
(
    c_customer_sk             bigint,
    c_customer_id             string,
    c_current_cdemo_sk        bigint,
    c_current_hdemo_sk        bigint,
    c_current_addr_sk         bigint,
    c_first_shipto_date_sk    bigint,
    c_first_sales_date_sk     bigint,
    c_salutation              string,
    c_first_name              string,
    c_last_name               string,
    c_preferred_cust_flag     string,
    c_birth_day               int,
    c_birth_month             int,
    c_birth_year              int,
    c_birth_country           string,
    c_login                   string,
    c_email_address           string,
    c_last_review_date        string
)
stored as parquet
location 'gs://?';

create external table tpcds_parquet.store
(
    s_store_sk                bigint,
    s_store_id                string,
    s_rec_start_date          string,
    s_rec_end_date            string,
    s_closed_date_sk          bigint,
    s_store_name              string,
    s_number_employees        int,
    s_floor_space             int,
    s_hours                   string,
    s_manager                 string,
    s_market_id               int,
    s_geography_class         string,
    s_market_desc             string,
    s_market_manager          string,
    s_division_id             int,
    s_division_name           string,
    s_company_id              int,
    s_company_name            string,
    s_street_number           string,
    s_street_name             string,
    s_street_type             string,
    s_suite_number            string,
    s_city                    string,
    s_county                  string,
    s_state                   string,
    s_zip                     string,
    s_country                 string,
    s_gmt_offset              double,
    s_tax_precentage          double                  
)
stored as parquet
location 'gs://?';

create external table tpcds_parquet.store_returns
(
    sr_returned_date_sk       bigint,
    sr_return_time_sk         bigint,
    sr_item_sk                bigint,
    sr_customer_sk            bigint,
    sr_cdemo_sk               bigint,
    sr_hdemo_sk               bigint,
    sr_addr_sk                bigint,
    sr_store_sk               bigint,
    sr_reason_sk              bigint,
    sr_ticket_number          bigint,
    sr_return_quantity        int,
    sr_return_amt             double,
    sr_return_tax             double,
    sr_return_amt_inc_tax     double,
    sr_fee                    double,
    sr_return_ship_cost       double,
    sr_refunded_cash          double,
    sr_reversed_charge        double,
    sr_store_credit           double,
    sr_net_loss               double             
)
stored as parquet
location 'gs://?';


create external table tpcds_parquet.date_dim
(
    d_date_sk                 bigint,
    d_date_id                 string,
    d_date                    string,
    d_month_seq               int,
    d_week_seq                int,
    d_quarter_seq             int,
    d_year                    int,
    d_dow                     int,
    d_moy                     int,
    d_dom                     int,
    d_qoy                     int,
    d_fy_year                 int,
    d_fy_quarter_seq          int,
    d_fy_week_seq             int,
    d_day_name                string,
    d_quarter_name            string,
    d_holiday                 string,
    d_weekend                 string,
    d_following_holiday       string,
    d_first_dom               int,
    d_last_dom                int,
    d_same_day_ly             int,
    d_same_day_lq             int,
    d_current_day             string,
    d_current_week            string,
    d_current_month           string,
    d_current_quarter         string,
    d_current_year            string 
)
stored as parquet
location 'gs://?';
```

**Comment by kokosing** [comment]
I will take care of this.

**Comment by kokosing** [comment]
This kind of correlated subquery is not yet supported, but if you change this subquery to (what was originally in TPCDS):

```
(SELECT avg(return_limit)
FROM high_return
WHERE  ctr1.ctr_store_sk = high_return.ctr_store_sk)
```

it will start to work. Instead of `avg` you could use `arbitrary`. 

Anyway, it would be still nice to have this query passing without any change, I will try work on this, but since correlated subqueries is not at the highest priority at the moment it is hard to tell when it will be fixed.

**Comment by georgewfraser** [comment]
Ironically I was refactoring this query so that bigquery would run it---I'm trying to create a version of tpcds that will run unmodified on 4 warehouses. 

Presto wasn't the only one that rejected this subquery, and it can be easily fixed by converting to a join. It would just be nice if it produced a better error message that indicated "this is an evil subquery and I can't figure out how to convert it to a join, you do it".

**Comment by martint** [comment]
This is related to https://github.com/prestodb/presto/issues/6580.

**Comment by kokosing** [comment]
Here is a fix for this #8435

**Comment by kokosing** [comment]
Fixed with https://github.com/prestodb/presto/commit/1ee30393c90899472a7b19e8017378a807f4191f
